circularBuffer.cpp

```diff
@@ -0,0 +1,43 @@
+#include <iostream>
+#include "circularBuffer.hpp"
+
+template <class T> T      circularBuffer<T>::buffer[BufferSize] = {};
+template <class T> size_t circularBuffer<T>::head =                0;
+template <class T> size_t circularBuffer<T>::tail =                0;
+template <class T> bool   circularBuffer<T>::full =            false;
+
+template <class T> void circularBuffer<T>::put_overwrite(T item)
+{
+  buffer[head] = item;
+  if (true == full)
+  {
+    tail = (tail + 1) % BufferSize;
+  }
+  head = (head + 1) % BufferSize;
+  full = (head == tail);
+  return;  
+}
+
+template <class T> void circularBuffer<T>::put_discard(T item)
+{
+  if (false == full)
+  {
+    buffer[head] = item;
+    head = (head + 1) % BufferSize;
+    full = (head == tail);
+  }
+  return;
+}
+
+template <class T> T circularBuffer<T>::get()
```

Comment:
How would you change this method to not return any data when empty?

Reply:
Could be done by returning a bool value. True if buffer was not empty, false if buffer is empty. 

template <class T, uint32_t BufferSize> bool circularBuffer<T, BufferSize>::get(T &item)

Call by reference used for the item.

Reply:
👍